proposals/0370-modifiers.rst

```diff
@@ -122,39 +126,136 @@ Proposed Change Specification
     have a known type if declared with a type signature. Alternatively, the
     modifier may have a top-level type signature.
 
-12. A modifier of type ``Multiplicity`` changes the multiplicity of the following arrow,
-    or following pattern-bound variable of a lambda,
-    or preceding record field.
-    Multiple modifiers of type ``Multiplicity`` on the same arrow are not allowed.
-    Any other use of a modifier is an error.
-
-13. ``-XLinearTypes`` implies ``-XModifiers``.
-
-14. Future modifiers will be put *before* the element they modify. Alternatively,
+12. Future modifiers will be put *before* the element they modify. Alternatively,
     a modifier can be put directly before a syntactic closer or separator, such
     as ``;`` or ``where`` or ``)``.
 
-15. Modifiers with an unknown meaning produce a warning, controlled by
-    ``-Wunknown-modifiers``. They are otherwise ignored. (However, in order to
-    know that a modifier is unknown, it still must be parsed, renamed, and type-checked.)
+13. Modifiers of unknown kind produce an error.
+
+14. Modifiers of known kind but with an unknown meaning produce a warning,
+    controlled by ``-Wunknown-modifiers``. They are otherwise ignored. (However,
+    in order to know that a modifier is unknown, it still must be parsed,
+    renamed, and type-checked.)
+
+LinearTypes
+-----------
+With ``-XLinearTypes``:
+
+* A modifier of type ``Multiplicity`` changes the multiplicity of the following
+  arrow, or following pattern-bound variable of a lambda, or following let or
+  where binding, or preceding record field. Multiple modifiers of type
+  ``Multiplicity`` on the same arrow are not allowed. Any other use of a
+  modifier still has no meaning.
+
+* The ``%1`` modifier is handled as a special case. It's renamed (and so
+  typechecked) the same as ``%One``, even if it appears somewhere that linear
+  modifiers aren't expected. If a user does want the modifier ``1 :: Nat``, they
+  can write it as ``%01``. (Requires ``-XDataKinds``.)
+
+* The linear arrow ``a ⊸ b`` has the same meaning as ```a %1 -> b``. Other
+  modifiers are accepted: ``a %Matchable ⊸ b`` has the same meaning as
+  ``a %Matchable %1 -> b``.
+
+With ``-XNoLinearTypes``, the ``%1`` modifier is not special. It refers to the
+type ``1 :: Nat`` and requires ``-XDataKinds``.
+
+``-XLinearTypes`` implies ``-XModifiers``. But the latter can be explicitly
+disabled with ``-XLinearTypes -XNoModifiers``. This introduces backwards
+compatible behavior:
+
+* Only ``Multiplicity`` modifiers are permitted, and only in the places they're
+  recognized. Any use of a modifier is an error.
+
+* The kind of a modifier is determined by checking for ``Multiplicity``, not
```

Comment:
Ah, can you elaborate on how it's not? (I see it should have been "any other use" on 167, but I guess that's not what you mean.)

I'm also not sure I see the connection. I think that if "linear types without modifiers" isn't backwards compatible, that reduces the motivation to have special rules for that case. (As opposed to, say, "without modifiers, linear types has no effect simply because you can't access any of the behavior changes it implies".) But I don't see why it reduces the motivation to have `-XLinearTypes` imply `-XModifiers` by default. If anything I think it increases that motivation.

Reply:
I can write `Int %m -> Int` today, and GHC will infer `m :: Multiplicity`. That would not be true with the more general `-XModifiers` approach.

Reply:
Right, it might not have been clear: this section is specifically intended as "if the user wants backwards compatibility, here's how they can get it". So (according to the current state of this PR):

* With linear types and modifiers both enabled, we use synthesis-not-checking, and `Int %m -> Int` is an error.
* With linear types enabled and modifiers disabled, we go back to checking-not-synthesis, and `Int %m -> Int` is accepted. My intent is that this case is backwards compatible. I lean towards not keeping this around permanently, if we have it at all.
* To unlock that case, the user needs to write `-XLinearTypes -XNoModifiers`, which is admittedly a bit arcane.